# bioCC
Fast advanced correlation calculator for basics bioinformatics file formats.
It computes Pearson’s and signal’s  correlation coefficients for densities, coverages and features. 
Program allows to obtain correlation coefficients for the whole genome, for each chromosome separately and for predefined regions inside chromosomes.

**bioCC** is designed to treat a bunch of files at once.

It runs on the command line under Windows, Linux and Mac OS X.<br>
Non-numeric option values are case insensitive.

## Usage
```
bioCC [options] -g|--gen <file> file0 file1 ...
bioCC [options] -g|--gen <file> -l|--list <file>
```

## Help
```
Input:
  -a|--align            input beds are an alignment
  -d|--dupl <OFF|ON>    accept duplicate reads. For alignments only (ON)
  -g|--gen <file>       genome size file, or genome file name, or directory contained genome
  --gap-len <int>       minimal length of undefined nucleotides region in genome
                        which is declared as a gap. Ignored for genome size file (1000)
  -l|--list <file>      list of input files.
                        First (primary) file in list is comparing with others (secondaries)
Processing:
  -c|--chr <chars>      treat stated chromosome only (all)
  -r|--cc <P|S|PS>      correlation coefficient: P – Pearson, S - signal, PS - both (P)
  -s|--space <int>      resolution: span in bps in which reads will be counted to define a density.
                        For alignments only (50)
  -T|--total <ADD|ONLY> output total coefficients: ADD - in addition, ONLY - solely (NONE)
Regions processing:
  -f|--fbed <file>      'template' bed file which features defines compared regions.
                        Ignored for ordinary beds
  -e|--exp-len <int>    length of expanding features in first ordinary bed file
                        or in 'template' bed file (0)
  -b|--bin-width <float>width of bins of regions histogram (0)
  --sort <RGN|CC>       output coefficients for each region, sorted by:
                        RGN - regions, CC - coefficients [NONE]
  --norm <OFF|ON>       normalize regions before calculation [ON]
Output:
  --alarm               output features ambiguities, if they exist. Ignored for wigs (OFF)
  --lac                 laconic output
  --stat                output features ambiguities statistics, if they exist.
                        Ignored for wigs
  -o|--out              duplicate standard output to bioCC_out.txt file
Other:
  -t|--time             output run time
  -v|--version          print program's version and quit
  -h|--help             print usage information and quit
```

## Details

### Input data types
Alignment read’s densities should be presented by [BED](https://genome.ucsc.edu/FAQ/FAQformat.html#format1) format.<br>
Read’s coverages should be presented by [WIG](https://genome.ucsc.edu/goldenpath/help/wiggle.html) format.<br>
Features are also presented by [BED](https://genome.ucsc.edu/FAQ/FAQformat.html#format1) format.

Since bed file as a result of peak calling process and bed file represents alignment have common required fields formally, but differ in using and interpretation, to be more clear we would call first type of bed *ordinary bed*, and second – *alignment bed*.

For more information about read’s density see option ```-s|--space```.

All input files should have chromosomes items (features, data lines) been clustering together and ordering by increase.<br>
Wig files generated by [MACS](http://liulab.dfci.harvard.edu/MACS/00README.html) or [PeakRanger](http://ranger.sourceforge.net/manual1.18.html) always meet these requirements. Opposite, bed files usually have messy initialization.  The simplest way to prepare bed files is to sort them (though the order of chromosomes themselves does not matter).<br>
It concerns ‘template’ bed file as well (see ```-f|--fbed``` option).

**bioCC** recognizes format automatically by filename extention. To distinguish ordinary beds and alignments a special option is provided.

Zipped input files (.gz) are accepted as well.

### Input data order
Input data can be provided as program parameters or as a file, contained file names. In both cases the first file in a list – *primary* – is compared to the others – *secondary* – by turns.
Be careful by using standard naming conventions like *abc?.bed*, *a\*.bed*.  You should be sure that first lexically recognized file is what you really want to be primary, and that all the others are really necessary to be compare.
For more details about list of file names as a file see option ```-l|--list```.

### Options description
```-a|--align```<br>
Indicate that input bed files are alignments, so density correlation would be performed.<br>
Since alignment may be considered as an “instance” of ordinary bed, in idea both cases should get the same result. It is true when the alignments as well as the ordinary beds have no *ambiguous* reads/features (see ```-–alarm``` option for definition). But in practice they almost always have, and these ambiguities are resolved by different way.  Consequently the results can be dramatically different. This is a reason of this option.<br>
In addition ordinary beds are correlated by a special extremely fast algorithm.<br>
If alignments are treated without this option, **bioCC** will print a warning message.<br>
If ordinary bed files have features of different sizes and are treated with this option, **bioCC** will print a cancel message and complete.

```-d|--dupl <OFF|ON>```<br>
Accept or deny duplicated reads to participate in density calculation.<br>
This option is topical for alignments only.<br>
Default: ON

```-g|--gen <name>```<br>
Reference genome size file, or reference genome library, or single reference nucleotide sequence.<br>
Genome library is a directory contained nucleotide sequences for each chromosome in [FASTA]() format.<br>
The difference between genome size file and genome library/file is that in the last case all https://en.wikipedia.org/wiki/FASTA_formatthe undefined regions (gaps), will be excluded from correlation process.<br>
Undefined regions are regions with only ambiguous reference characters ‘N’ in them.<br>
The minimal length of accounting gaps is managed by ```--gap-len``` option.<br>
For example, chromosome 1 from mm9 library contains 14 regions, separated by gaps with length more then 400 bps, and 10 regions, separated by gaps with length more then 1000.<br>
Indicating genome library has the same effect as ```-f|--fbed``` option, where ‘template’ is a set of defined regions.

It may signify in some special cases. If you are comparing control data, “inputs” (f.e. to define how really stochastic these data are), excluding undefined regions reduces Pearson coefficient up to 18%.<br>
That is a measure of “ordering” which these regions are introducing into irregular data (“chaos”).

You can obtain genome library in UCSC or in ensemble storage. In the second case please copy genomic sequences with the same masked type only, f.e. unmasked (‘dna'), since program does not recognise mask’s types.<br>
Zipped .fa files can be mix with unzipped.<br>
The single pointed FASTA file has the same effect as ```-c|--chr``` option.<br>
Scanning the whole genome library consumes some time. For this reason bioCC scans only chromosomes which  are treating (included in compared files or pointed by -c|--chr option). Moreover, program makes  it just ones, and writes discovered defined regions into short files named like *chrZ.region*, located in the same directory. If on the next call **bioCC** has the same value of ```--gap-len``` option, it reads these files by requery.<br>
In any case genome size file always fetches the fastest way.<br>
This option is required.

```--gap-len <int>```<br>
Minimal length of undefined nucleotides region which is taken as a gap. For more details see ```-g|--gen``` option.<br>
Ignored for genome size file, pointed by ```-g|--gen```.<br>
Default: 1000

```-l|--list <file>```<br>
External list of input files: a plain text file, on which each input file’s name is placed on separate line.<br>
Lines beginning with ‘#’ are considered as comments and will be skipped.<br>
Empty lines are also allowed.<br>
This option abolishes input files as parameters.

```-c|--chr <chars>```<br>
Treat stated chromosome only. Samples of option’s value: 1, 20, X.<br>
Reduces run time on 1.5-20 times depends of how far this chromosome is placed in an input alignment or wig.<br>
For ordinary beds it has no time-improvement effect: any result appears in a moment.<br>
Default: all

```-r|--cc <P|S|PS>```<br>
Correlation coefficient: P – Pearson, S – signal, PS – both.<br>
For more details see ‘Pearson and signal correlation’ section.<br>
Default: Pearson

```-s|--space <int>```<br>
The resolution for calculating reads (tags) density.  This is a space on which each chromosome is cutting and for which reads are counting. Read is considered to belong to certain space if its centre is inside this space. So, each read is counted just one time. As a result we obtain a read density distribution.<br>
This option is topical for alignments only, since for wiggles resolution is a part of data.

It needs to clarify the last point.
**bioCC** takes the resolution of primary wiggle as the basic for treatment. If secondary wiggle’s resolution is differ, it will be result to basic. For example, if primary resolution is 10, and secondary is 1, secondary data will be amalgamate to reduce resolution to 10. Contrariwise, if secondary resolution  is 100, each single data will be repeated in memory 9 times to level resolutions.
Both cases are not optimal: in first case we lose the time (for reading redundant data), in second – the memory (for keeping recurrent data).
The best way is to use wiggles with the same resolutions.

Default: 10.

```-T|--total <ADD|ONLY>```<br>
If set, output total coefficients: ADD – in addition to chromosomes results, ONLY – solely.<br>
If single chromosome is pointed, total result is omitted because of its equivalence.

```-f|--fbed <file>```<br>
'Template' ordinary bed file with features that defines compared regions.<br>
In practice the most interesting case is comparing detected peaks across the distributions, for what this option is mainly constructed.<br>
If options ```-b|--bin-width``` and ```--fr``` are not defined, only total coefficients for the all regions are printed.<br>
If this option is established, predefined regions from genome library are ignored (assuming that features from peak callers are always setting in defined genome regions).<br>
This option is ignored for ordinary input bed files.

```-e|--exp-len <int>```<br>
Length of expanding features in 'template' bed file or in primary ordinary bed file.<br>
If set, all the features from 'template' or primary bed file are be-expanding before processing: *start* positions are decreased for this value, *end* positions are increased.<br>
If expanded features become intersected, they are joined. See options ```--alarm``` and ```–-stat``` for more information.<br>
This option is constructed for the special cases, when you need to compare alignment (density or coverage), or ordinary bed file from peak caller, with the file contained “real” peaks.<br>
For example, “real” peaks can be the somehow predefined TFBS. Real TFBS usually have much more less length then length of relevant enriched regions or recovered peaks (8-20 versus 500-2000 pbs). Direct comparison “real” bed file and bed file after peak caller gives an inadequate low correlation coefficient even in case of complete correspondence of “real” and recovered peaks. If you expand primary TFBS length to the mean length of recovered peaks, you can obtain the true coefficient.<br>
Direct comparison of alignment’s regions, which are the “real” TFBS established in a ‘template’, also will give a rather wrong result since only the narrow fragments of the enriched regions will be correlated.  In this situation expanding ‘template’ regions gives you a way to reach the true coefficient as well.<br>
Default: 0

```-b|--bin-width <float>```<br>
If more then 0, force to consolidate coefficients calculated for regions into histogram, and print them.
Histogram shows count of coefficients (frequency) within some value ranges (bins). It is printed as a list of pairs *\<bin upper bound\> –  \<count in bin\>*. Negative coefficients have been turning to absolute during consolidation.<br>
Option’s value defines width of bin as a part of 1. Not all the bins are printed: empty bins at the edges are skipped.<br>
For example, if –b value is 0.1, and all coefficients are placing in the range 0.65 to 0.85, only three bins [0.6-] 0.7, [0.7-] 0.8, [0.8-] 0.9 would be printed.<br>
This option is topical only with option ```-f|--fbed``` or for genome library.<br>
Default: 0.

```--sort <RGN|CC>```<br>
If set, force to output coefficients calculated for each region as a list of pairs *\<number-of-region\> – \<coefficient\>*. Value RGN prescribes list to be sorted by region’s number, CC – by coefficient.<br>
If both of the coefficients are declared, list is sorted by Pearson.<br>
First region number is 1.<br>
This option is topical only with option ```-f|--fbed``` or for genome library.<br>

```--rn <OFF|ON>```<br>
Normalize regions before calculation. <br>
Normalization means levelling distribution’s patterns by their maximal values across regions.<br>
In spite of each pair of regions will be always compared correctly, the common result for all the regions may been unfairly falling down. It occurs when patterns levels are appreciably differ from region to region.<br>
This option is assigned to eliminate such effect.<br>
It is topical only with option ```-f|--fbed``` or for genome library.<br>
Default: ON

```--alarm```<br>
Output ambiguities, if they exist.<br>
When we are speaking about bed files, there are some situations which can be normal or may be not – that depends of bed’s destination. For example, duplicated and crossed features are normal for alignments, but rather unusual for ordinary beds. On the other hand, features with different length are typical for ordinary beds, but have been denied for alignments. All these situations we call a conditional term ambiguities.<br>
It concerns to ‘template’ bed as well.<br>
**bioCC** treats ambiguities the most natural way. For ordinary bed it merges crossed or adjacent features, and ignores submerging and duplicated features. For alignment it accepts duplicated reads by default (in fact, all the ambiguities are permitted for the alignment).<br>
But in some circumstances you need to be aware about these facts. There are two ways to know about them: detailed and summary.<br>
This option provides a detailed way. If it is pointed, information about type of ambiguity, number of line where it occurs, and resulting treatment would be printed each time when it happens.<br>
Summary way is managed by option ```–-stat```.<br>
Duplicated reads are not printed this way since they are considered as normal for alignment, but they are reported in summary way as well.<br>
This option ignored for input wigs.

```--lac```<br>
Laconic output. This option minimizes the output as possible to remain unambiguous. Mainly it comes in handy by using in batch file.<br>
Do not lock alarm and statistics output.

```--stat```<br>
Output features ambiguities statistics, if they exist. Prints number of all recognized certain type ambiguities, and appropriate treatment.<br>
In particular, this is a simple way to know the number of duplicated reads in alignments.<br>
This option also forces to output initial number of features/reads, and number of features/reads after ambiguity treatment, if they are different.<br>
For more details about ambiguities see ```--alarm``` option.<br>
This option ignored for input wigs.

```-o|--out```<br>
Duplicate standard output to bioCC_out.txt file.<br>
It is analogue of **tee** Linux command and rather is useful by calling under Windows .


## Pearson and signal correlation
Pearson’s method is right because of linear nature of comparing data.<br>
While we consider coverage/density as the data distributed along regular dimension (scaled chromosome’s length), signal’s methods is appropriate as well.<br>
The only difference between them is the subtractive mean in covariance on Pearson’s coefficient.

What consequences it takes, and which coefficient is better to choose?

A. Coverage/density distributions.<br>
To be more clear there are 3 illustrations of pair of functions which are quite near to the fragments of real coverages.<br>
Both of coefficients demonstrate value’s normalization independence (![fig 1,3](https://github.com/fnaumenko/bioCC/tree/master/pict/Signal-Pearson.png)).<br>
Signal’s method is a bit more sensible. But if we are interesting of measure of the degree of distribution patterns (shapes) only, this method becomes inappropriate when some distribution have non-zero background (*DC offset* in terms of signal function), as it shows in ![fig 2](https://github.com/fnaumenko/bioCC/tree/master/pict/Signal-Pearson.png)<br>
Briefly, Pearson’s method is mainly recommended for the distributions comparison.<br>
But if background’s level is considered part of the measure of similarity, in this case signal’s method would be preferable (it is true if resolution is sufficiently low to average background’s irregularity).

B.  Bed features.
Bed features can be accounted as discrete function accepted one of two values: 0 or 1. In that case signal’s method becomes inappropriate due to obvious reason: it counts intersections only. For example, by comparison two mutually complementary beds (while function1 have zero value every time when function2 have non-zero and vice versa), signal coefficient would be 0. Though the right answer is -1.<br>
So, to features only Pearson’s method is appropriate.
